mw_transform_jq.go

```diff
@@ -0,0 +1,115 @@
+package main
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"github.com/Sirupsen/logrus"
+	"github.com/mitchellh/mapstructure"
+	"io/ioutil"
+	"net/http"
+)
+
+type TransformJQMiddleware struct {
+	*BaseMiddleware
+}
+
+type JQTransformOptions struct {
+	OutputHeaders map[string]string `mapstructure:"output_headers"`
+	OutputVars    map[string]string `mapstructure:"output_vars"`
+}
+
+func (t *TransformJQMiddleware) GetName() string {
+	return "TransformJQMiddleware"
+}
+
+func (t *TransformJQMiddleware) IsEnabledForSpec() bool {
+	for _, version := range t.Spec.VersionData.Versions {
+		if len(version.ExtendedPaths.TransformJQ) > 0 {
+			return true
+		}
+	}
+	return false
+}
+
+// ProcessRequest will run any checks on the request on the way through the system, return an error to have the chain fail
+func (t *TransformJQMiddleware) ProcessRequest(w http.ResponseWriter, r *http.Request, _ interface{}) (error, int) {
+	_, versionPaths, _, _ := t.Spec.GetVersionData(r)
+	found, meta := t.Spec.CheckSpecMatchesStatus(r, versionPaths, TransformedJQ)
+	if !found {
+		return nil, 200
+	}
+	err := transformJQBody(r, meta.(*TransformJQSpec), t.Spec.EnableContextVars)
+	if err != nil {
+		log.WithFields(logrus.Fields{
+			"prefix":      "inbound-transform-jq",
+			"server_name": t.Spec.Proxy.TargetURL,
+			"api_id":      t.Spec.APIID,
+			"path":        r.URL.Path,
+		}).Error(err)
+		return err, 415
+	}
+	return nil, 200
+}
+
+func transformJQBody(r *http.Request, t *TransformJQSpec, contextVars bool) error {
+	// Read the body:
+	defer r.Body.Close()
+	body, err := ioutil.ReadAll(r.Body)
+	if err != nil {
+		return err
+	}
+
+	var bodyObj map[string]interface{}
+	if err := json.Unmarshal(body, &bodyObj); err != nil {
+		return err
+	}
+
+	jqObj := map[string]interface{}{
+		"body":       bodyObj,
+		"reqContext": ctxGetData(r),
+	}
+
+	if err = t.JQFilter.Handle(jqObj); err != nil {
+		return err
+	}
+
+	// First element is the transformed body
+	if t.JQFilter.Next() {
```

Comment:
do `if !x { return errors.New(...) }` to simplify the flow and unindent the other code.